File: sdk/javascript/smoke-test/test_smoke_composite.ts
import { PaymentClient, MerchantAuthenticationClient, types, IntegrationError, ConnectorError } from "hyperswitch-prism";
import * as fs from "fs";
import * as path from "path";

const {
  PaymentStatus,
  Connector,
} = types;

const { RequestConfig, Environment } = types;

const defaults = RequestConfig.create({});

interface Credentials {
  stripe?: { apiKey?: { value: string }; api_key?: { value: string } } | Array<{ apiKey?: { value: string }; api_key?: { value: string } }>;
  paypal?: { clientId?: { value: string }; client_id?: { value: string }; clientSecret?: { value: string }; client_secret?: { value: string } } | Array<{ clientId?: { value: string }; client_id?: { value: string }; clientSecret?: { value: string }; client_secret?: { value: string } }>;
}

function loadCredentials(credsFile: string): Credentials {
  if (!fs.existsSync(credsFile)) return {};
  return JSON.parse(fs.readFileSync(credsFile, "utf-8"));
}

function getStripeApiKey(credentials: Credentials): string | null {
  if (!credentials.stripe) return null;
  const stripeCreds = Array.isArray(credentials.stripe) ? credentials.stripe[0] : credentials.stripe;
  return stripeCreds.apiKey?.value || stripeCreds.api_key?.value || null;
}

function getPayPalCredentials(credentials: Credentials): { clientId: string; clientSecret: string } | null {
  if (!credentials.paypal) return null;
  const paypalCreds = Array.isArray(credentials.paypal) ? credentials.paypal[0] : credentials.paypal;
  const clientId = paypalCreds.clientId?.value || paypalCreds.client_id?.value;
  const clientSecret = paypalCreds.clientSecret?.value || paypalCreds.client_secret?.value;
  if (clientId && clientSecret) {
    return { clientId, clientSecret };
  }
  return null;
}

async function testPaypalAuthorize(credsFile: string): Promise<boolean> {
  console.log("\n[PayPal Authorize]");

  if (!fs.existsSync(credsFile)) {
    console.log("  SKIPPED: creds.json not found");
    return true;
  }
  const credentials = loadCredentials(credsFile);
  const paypalCreds = getPayPalCredentials(credentials);

  if (!paypalCreds) {
    console.log("  SKIPPED: No PayPal credentials in creds.json");
    return true;
  }

  console.log(`  Using client_id: ${paypalCreds.clientId.substring(0, 10)}...`);

  const paypalConfig: types.IConnectorConfig = {
    options: { environment: Environment.SANDBOX },
    connectorConfig: {
      paypal: {
        clientId: { value: paypalCreds.clientId },
        clientSecret: { value: paypalCreds.clientSecret },
      },
    },
  };

  const authClient = new MerchantAuthenticationClient(paypalConfig, defaults);
  const paymentClient = new PaymentClient(paypalConfig, defaults);

  // Step 1: Create access token
  console.log("  Step 1: Creating access token...");
  let accessTokenValue: string | null = null;
  let tokenTypeValue: string | null = null;
  let expiresInSeconds: number = 3600;

  try {
    const accessTokenResponse = await authClient.createServerAuthenticationToken({
      merchantAccessTokenId: "paypal_token_" + Date.now(),
      connector: Connector.PAYPAL,
      testMode: true,
    });

    if (accessTokenResponse.accessToken?.value) {
      accessTokenValue = accessTokenResponse.accessToken.value;
      tokenTypeValue = accessTokenResponse.tokenType ?? "Bearer";
      expiresInSeconds = Number(accessTokenResponse.expiresInSeconds) || 3600;
      console.log("  Access token received");
    } else {
      console.log("  No access token in response");
      return true;
    }
  } catch (e: any) {
    if (e instanceof IntegrationError) {
      console.log(`  IntegrationError: ${e.message} (code=${e.errorCode}, action=${e.suggestedAction}, doc=${e.docUrl})`);
      return true;
    }
    if (e instanceof ConnectorError) {
      console.log(`  ConnectorError: ${e.message} (code=${e.errorCode}, http=${e.httpStatusCode})`);
      return true;
    }
    console.log(`  Error creating access token: ${e.message}`);
    return true;
  }

  // Step 2: Authorize with access token
  console.log("  Step 2: Authorizing with access token...");
  const authorizeRequest: types.PaymentServiceAuthorizeRequest = {
    merchantTransactionId: "paypal_authorize_" + Date.now(),
    amount: { minorAmount: 1000, currency: types.Currency.USD },
    captureMethod: types.CaptureMethod.AUTOMATIC,
    paymentMethod: {
      card: {
        cardNumber: { value: "4111111111111111" },
        cardExpMonth: { value: "12" },
        cardExpYear: { value: "2050" },
        cardCvc: { value: "123" },
        cardHolderName: { value: "Test User" },
      },
    },
    state: {
      accessToken: {
        token: { value: accessTokenValue },
        tokenType: tokenTypeValue,
        expiresInSeconds,
      },
    },
    address: { billingAddress: {} },
    authType: types.AuthenticationType.NO_THREE_DS,
    returnUrl: "https://example.com/return",
    orderDetails: [],
  };

  try {
    const response = await paymentClient.authorize(authorizeRequest);

    if (response.status === PaymentStatus.CHARGED) {
      console.log("  PASSED: Payment charged");
      return true;
    } else {
      console.log(`  FAILED: Expected CHARGED, got ${response.status}`);
      return false;
    }
  } catch (e: any) {
    if (e instanceof IntegrationError) {
      console.log(`  IntegrationError: ${e.message} (code=${e.errorCode}, action=${e.suggestedAction}, doc=${e.docUrl})`);
      return true;
    }
    if (e instanceof ConnectorError) {
      console.log(`  ConnectorError: ${e.message} (code=${e.errorCode}, http=${e.httpStatusCode})`);
      return true;
    }
    console.error("  Error:", e.message || e);
    return false;
  }
}

function parseArgs(): { credsFile: string } {
  const args = process.argv.slice(2);
  let credsFile = "creds.json";

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "--creds-file" && i + 1 < args.length) {
      credsFile = args[++i];
    }
  }

  // Resolve relative paths from cwd
  if (!path.isAbsolute(credsFile)) {
    credsFile = path.join(process.cwd(), credsFile);
  }

  return { credsFile };
}

/**
 * IntegrationError path: missing required field (amount) must throw IntegrationError
 * and must NOT reach the connector.
 */
async function testStripeIntegrationError(credsFile: string): Promise<boolean> {
  console.log("\n[Stripe IntegrationError — missing amount]");

  if (!fs.existsSync(credsFile)) {
    console.log("  FAILED: creds.json not found");
    return false;
  }
  const credentials = loadCredentials(credsFile);
  const apiKey = getStripeApiKey(credentials);

  if (!apiKey) {
    console.log("  FAILED: No Stripe API key in creds.json");
    return false;
  }

  const stripeConfig: types.ConnectorConfig = {
    options: { environment: Environment.SANDBOX },
    connectorConfig: { stripe: { apiKey: { value: apiKey } } },
  };

  const paymentClient = new PaymentClient(stripeConfig, defaults);

  // amount intentionally omitted
  const authorizeRequest: types.PaymentServiceAuthorizeRequest = {
    merchantTransactionId: "stripe_missing_amount_" + Date.now(),
    paymentMethod: {
      card: {
        cardNumber: { value: "4111111111111111" },
        cardExpMonth: { value: "12" },
        cardExpYear: { value: "2050" },
        cardCvc: { value: "123" },
        cardHolderName: { value: "Test User" },
      },
    },
    captureMethod: types.CaptureMethod.AUTOMATIC,
    address: { billingAddress: {} },
    authType: types.AuthenticationType.NO_THREE_DS,
    returnUrl: "https://example.com/return",
    orderDetails: [],
  };

  try {
    await paymentClient.authorize(authorizeRequest);
    console.log("  FAILED: Expected IntegrationError but call succeeded — request should have been rejected before the HTTP call");
    return false;
  } catch (e: any) {
    if (e instanceof IntegrationError) {
      console.log(`  PASSED: IntegrationError (expected): ${e.message} (code=${e.errorCode})`);
      return true;
    }
    if (e instanceof ConnectorError) {
      console.log(`  FAILED: Got ConnectorError instead of IntegrationError: ${e.message} (code=${e.errorCode}, http=${e.httpStatusCode})`);
      return false;
    }
    console.error("  FAILED: Unexpected error:", e.message || e);
    return false;
  }
}

/**
 * ConnectorError path: request is valid but card is known to be declined by Stripe.
 * Must throw ConnectorError, not IntegrationError.
 */
async function testStripeConnectorError(credsFile: string): Promise<boolean> {
  console.log("\n[Stripe ConnectorError — declined card]");

  if (!fs.existsSync(credsFile)) {
    console.log("  FAILED: creds.json not found");
    return false;
  }
  const credentials = loadCredentials(credsFile);
  const apiKey = getStripeApiKey(credentials);

  if (!apiKey) {
    console.log("  FAILED: No Stripe API key in creds.json");
    return false;
  }

  const stripeConfig: types.ConnectorConfig = {
    options: { environment: Environment.SANDBOX },
    connectorConfig: { stripe: { apiKey: { value: apiKey } } },
  };

  const paymentClient = new PaymentClient(stripeConfig, defaults);

  const authorizeRequest: types.PaymentServiceAuthorizeRequest = {
    merchantTransactionId: "stripe_declined_" + Date.now(),
    amount: { minorAmount: 1000, currency: types.Currency.USD },
    captureMethod: types.CaptureMethod.AUTOMATIC,
    paymentMethod: {
      card: {
        cardNumber: { value: "4000000000000002" }, // Stripe generic decline test card
        cardExpMonth: { value: "12" },
        cardExpYear: { value: "2050" },
        cardCvc: { value: "123" },
        cardHolderName: { value: "Test User" },
      },
    },
    address: { billingAddress: {} },
    authType: types.AuthenticationType.NO_THREE_DS,
    returnUrl: "https://example.com/return",
    orderDetails: [],
  };

  try {
    await paymentClient.authorize(authorizeRequest);
    // Stripe should decline 4000000000000002 — if it doesn't, not our failure
    console.log("  PASSED: Card unexpectedly succeeded (sandbox may behave differently)");
    return true;
  } catch (e: any) {
    if (e instanceof ConnectorError) {
      console.log(`  PASSED: ConnectorError (expected): ${e.message} (code=${e.errorCode}, http=${e.httpStatusCode})`);
      return true;
    }
    if (e instanceof IntegrationError) {
      console.log(`  FAILED: Got IntegrationError instead of ConnectorError: ${e.message} (code=${e.errorCode})`);
      return false;
    }
    console.error("  FAILED: Unexpected error:", e.message || e);
    return false;
  }
}

async function testStripeAuthorize(credsFile: string): Promise<boolean> {
  console.log("\n[Stripe Authorize]");

  if (!fs.existsSync(credsFile)) {
    console.log("  FAILED: creds.json not found");
    return false;
  }
  const credentials = loadCredentials(credsFile);
  const apiKey = getStripeApiKey(credentials);

  if (!apiKey) {
    console.log("  FAILED: No Stripe API key in creds.json");
    return false;
  }

  const stripeConfig: types.ConnectorConfig = {
    options: { environment: Environment.SANDBOX },
    connectorConfig: { stripe: { apiKey: { value: apiKey } } },
  };

  const paymentClient = new PaymentClient(stripeConfig, defaults);

  const authorizeRequest: types.PaymentServiceAuthorizeRequest = {
    merchantTransactionId: "stripe_authorize_" + Date.now(),
    amount: { minorAmount: 1000, currency: types.Currency.USD },
    captureMethod: types.CaptureMethod.AUTOMATIC,
    paymentMethod: {
      card: {
        cardNumber: { value: "4111111111111111" },
        cardExpMonth: { value: "12" },
        cardExpYear: { value: "2050" },
        cardCvc: { value: "123" },
        cardHolderName: { value: "Test User" },
      },
    },
    address: { billingAddress: {} },
    authType: types.AuthenticationType.NO_THREE_DS,
    returnUrl: "https://example.com/return",
    orderDetails: []
  };

  try {
    const response = await paymentClient.authorize(authorizeRequest);

    if (response.status === PaymentStatus.CHARGED) {
      console.log("  PASSED: Payment charged");
      return true;
    } else {
      console.log(`  FAILED: Expected CHARGED, got ${response.status}`);
      return false;
    }
  } catch (e: any) {
    if (e instanceof IntegrationError) {
      console.log(`  IntegrationError: ${e.message} (code=${e.errorCode}, action=${e.suggestedAction}, doc=${e.docUrl})`);
      return true;
    }
    if (e instanceof ConnectorError) {
      console.log(`  ConnectorError: ${e.message} (code=${e.errorCode}, http=${e.httpStatusCode})`);
      return true;
    }
    console.error("  FAILED:", e.message || e);
    return false;
  }
}

async function main(): Promise<void> {
  const { credsFile } = parseArgs();
  let allPassed = true;

  const paypalPassed = await testPaypalAuthorize(credsFile);
  if (!paypalPassed) allPassed = false;

  const stripePassed = await testStripeAuthorize(credsFile);
  if (!stripePassed) allPassed = false;

  const stripeIntegrationErrorPassed = await testStripeIntegrationError(credsFile);
  if (!stripeIntegrationErrorPassed) allPassed = false;

  const stripeConnectorErrorPassed = await testStripeConnectorError(credsFile);
  if (!stripeConnectorErrorPassed) allPassed = false;

  console.log("\n" + "=".repeat(40));
  console.log(allPassed ? "PASSED" : "FAILED");
  process.exit(allPassed ? 0 : 1);
}

main().catch((e: any) => {
  console.error(e);
  process.exit(1);
});

File: sdk/python/smoke-test/test_smoke_composite.py
#!/usr/bin/env python3
"""
Composite smoke test for hyperswitch-payments SDK.

Tests:
  1. PayPal access token flow (create token + authorize)
  2. Stripe direct authorize flow

The test only passes when payments succeed (status == CHARGED).
CI will fail if any test fails.

Usage:
    python3 test_smoke_composite.py --creds-file creds.json
"""

import argparse
import asyncio
import json
import os
import sys
import time
from typing import Dict, Any, Optional, Tuple

# Add parent directory to path for imports when running directly
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

try:
    from payments import (
        PaymentClient,
        MerchantAuthenticationClient,
        ConnectorConfig,
        RequestConfig,
        Environment,
        Connector,
        Currency,
        CaptureMethod,
        AuthenticationType,
        PaymentStatus,
        MerchantAuthenticationServiceCreateAccessTokenRequest,
        PaymentServiceAuthorizeRequest,
        IntegrationError,
        ConnectorResponseTransformationError,
        PaymentAddress
    )
except ImportError as e:
    print(f"Error importing payments package: {e}")
    print(
        "Make sure the wheel is installed: pip install dist/hyperswitch_payments-*.whl"
    )
    sys.exit(1)


def parse_args() -> argparse.Namespace:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description="Composite smoke test for hyperswitch-payments SDK"
    )
    parser.add_argument(
        "--creds-file",
        default="creds.json",
        help="Path to connector credentials JSON file (default: creds.json)",
    )
    return parser.parse_args()


def load_credentials(creds_file: str) -> Dict[str, Any]:
    """Load connector credentials from JSON file."""
    if not os.path.exists(creds_file):
        return {}
    with open(creds_file, "r") as f:
        return json.load(f)


def get_stripe_api_key(credentials: Dict[str, Any]) -> Optional[str]:
    """Extract Stripe API key from credentials."""
    if not credentials.get("stripe"):
        return None
    stripe_creds = credentials["stripe"]
    if isinstance(stripe_creds, list):
        stripe_creds = stripe_creds[0]
    return stripe_creds.get("apiKey", {}).get("value") or stripe_creds.get(
        "api_key", {}
    ).get("value")


def get_paypal_credentials(credentials: Dict[str, Any]) -> Optional[Tuple[str, str]]:
    """Extract PayPal client_id and client_secret from credentials."""
    if not credentials.get("paypal"):
        return None
    paypal_creds = credentials["paypal"]
    if isinstance(paypal_creds, list):
        paypal_creds = paypal_creds[0]
    client_id = paypal_creds.get("clientId", {}).get("value") or paypal_creds.get(
        "client_id", {}
    ).get("value")
    client_secret = paypal_creds.get("clientSecret", {}).get(
        "value"
    ) or paypal_creds.get("client_secret", {}).get("value")
    if client_id and client_secret:
        return (client_id, client_secret)
    return None


async def test_paypal_authorize(creds_file: str) -> bool:
    """Test PayPal authorize flow with access token."""
    print("\n[PayPal Authorize]")

    if not os.path.exists(creds_file):
        print("  SKIPPED: creds.json not found")
        return True

    credentials = load_credentials(creds_file)
    paypal_creds = get_paypal_credentials(credentials)

    if not paypal_creds:
        print("  SKIPPED: No PayPal credentials in creds.json")
        return True

    client_id, client_secret = paypal_creds
    print(f"  Using client_id: {client_id[:10]}...")

    # Configure PayPal
    config = ConnectorConfig()
    config.options.environment = Environment.SANDBOX
    config.connector_config.paypal.client_id.value = client_id
    config.connector_config.paypal.client_secret.value = client_secret

    defaults = RequestConfig()
    auth_client = MerchantAuthenticationClient(config, defaults)
    payment_client = PaymentClient(config, defaults)

    # Step 1: Create access token
    print("  Step 1: Creating access token...")
    access_token_value = None
    token_type_value = "Bearer"
    expires_in_seconds = 3600

    try:
        access_token_request = MerchantAuthenticationServiceCreateAccessTokenRequest()
        access_token_request.merchant_access_token_id = (
            f"paypal_token_{int(time.time() * 1000)}"
        )
        access_token_request.connector = Connector.PAYPAL
        access_token_request.test_mode = True

        access_token_response = await auth_client.create_access_token(
            access_token_request
        )

        if (
            access_token_response.access_token
            and access_token_response.access_token.value
        ):
            access_token_value = access_token_response.access_token.value
            token_type_value = access_token_response.token_type or "Bearer"
            expires_in_seconds = access_token_response.expires_in_seconds or 3600
            print("  Access token received")
        else:
            print("  No access token in response")
            return True
    except Exception as e:
        print(f"  Error creating access token: {e}")
        return True

    # Step 2: Authorize with access token
    print("  Step 2: Authorizing with access token...")

    try:
        authorize_request = PaymentServiceAuthorizeRequest()
        authorize_request.merchant_transaction_id = (
            f"paypal_authorize_{int(time.time() * 1000)}"
        )
        authorize_request.amount.minor_amount = 1000
        authorize_request.amount.currency = Currency.USD
        authorize_request.capture_method = CaptureMethod.AUTOMATIC

        # Card details
        card = authorize_request.payment_method.card
        card.card_number.value = "4111111111111111"
        card.card_exp_month.value = "12"
        card.card_exp_year.value = "2050"
        card.card_cvc.value = "123"
        card.card_holder_name.value = "Test User"

        # Access token in state
        authorize_request.state.access_token.token.value = access_token_value
        authorize_request.state.access_token.token_type = token_type_value
        authorize_request.state.access_token.expires_in_seconds = expires_in_seconds

        authorize_request.auth_type = AuthenticationType.NO_THREE_DS
        authorize_request.address.CopyFrom(PaymentAddress())
        authorize_request.test_mode = True

        response = await payment_client.authorize(authorize_request)

        if response.status == PaymentStatus.CHARGED:
            print("  PASSED: Payment charged")
            return True
        else:
            print(f"  FAILED: Expected CHARGED, got {response.status}")
            return False
    except IntegrationError as e:
        print(f"  IntegrationError: {e.error_code} - {e.error_message}")
        return False
    except ConnectorResponseTransformationError as e:
        print(
            f"  ConnectorResponseTransformationError: {e.error_code} - {e.error_message}"
        )
        return False
    except Exception as e:
        print(f"  Error: {e}")
        return False


async def test_stripe_authorize(creds_file: str) -> bool:
    """Test Stripe authorize flow."""
    print("\n[Stripe Authorize]")

    if not os.path.exists(creds_file):
        print("  FAILED: creds.json not found")
        return False

    credentials = load_credentials(creds_file)
    api_key = get_stripe_api_key(credentials)

    if not api_key:
        print("  FAILED: No Stripe API key in creds.json")
        return False

    # Configure Stripe
    config = ConnectorConfig()
    config.options.environment = Environment.SANDBOX
    config.connector_config.stripe.api_key.value = api_key

    defaults = RequestConfig()
    payment_client = PaymentClient(config, defaults)

    try:
        authorize_request = PaymentServiceAuthorizeRequest()
        authorize_request.merchant_transaction_id = (
            f"stripe_authorize_{int(time.time() * 1000)}"
        )
        authorize_request.amount.minor_amount = 1000
        authorize_request.amount.currency = Currency.USD
        authorize_request.capture_method = CaptureMethod.AUTOMATIC

        # Card details
        card = authorize_request.payment_method.card
        card.card_number.value = "4111111111111111"
        card.card_exp_month.value = "12"
        card.card_exp_year.value = "2050"
        card.card_cvc.value = "123"
        card.card_holder_name.value = "Test User"

        authorize_request.auth_type = AuthenticationType.NO_THREE_DS
        authorize_request.address.CopyFrom(PaymentAddress())

        response = await payment_client.authorize(authorize_request)

        if response.status == PaymentStatus.CHARGED:
            print("  PASSED: Payment charged")
            return True
        else:
            print(f"  FAILED: Expected CHARGED, got {response.status}")
            return False
    except Exception as e:
        print(f"  FAILED: {e}")
        return False


async def main() -> int:
    """Run all composite tests."""
    args = parse_args()

    # Resolve relative paths from cwd
    creds_file = args.creds_file
    if not os.path.isabs(creds_file):
        creds_file = os.path.join(os.getcwd(), creds_file)

    all_passed = True

    # Run PayPal test (informational - can skip if no creds)
    paypal_passed = await test_paypal_authorize(creds_file)
    if not paypal_passed:
        all_passed = False

    # Run Stripe test (must pass)
    stripe_passed = await test_stripe_authorize(creds_file)
    if not stripe_passed:
        all_passed = False

    print("\n" + "=" * 40)
    if all_passed:
        print("PASSED")
        return 0
    else:
        print("FAILED")
        return 1


if __name__ == "__main__":
    exit_code = asyncio.run(main())
    sys.exit(exit_code)
